src/Middleware/RateLimiting/src/RateLimiterOptions.cs

```diff
@@ -44,5 +38,108 @@ public PartitionedRateLimiter<HttpContext> Limiter
     /// This status code will be set before <see cref="OnRejected"/> is called, so any status code set by
     /// <see cref="OnRejected"/> will "win" over this default.
     /// </remarks>
-    public int DefaultRejectionStatusCode { get; set; } = StatusCodes.Status503ServiceUnavailable;
+    public int RejectionStatusCode { get; set; } = StatusCodes.Status503ServiceUnavailable;
+
+    /// <summary>
+    /// Adds a new rate limiting policy with the given policyName.
+    /// </summary>
+    /// <param name="policyName">The name to be associated with the given <see cref="RateLimiter"/>.</param>
+    /// <param name="partitioner">Method called every time an Acquire or WaitAsync call is made to figure out what rate limiter to apply to the request.</param>
+    public RateLimiterOptions AddPolicy<TPartitionKey>(string policyName, Func<HttpContext, RateLimitPartition<TPartitionKey>> partitioner)
+    {
+        ArgumentNullException.ThrowIfNull(policyName);
+        ArgumentNullException.ThrowIfNull(partitioner);
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException($"There already exists a policy with the name {policyName}");
+        }
+
+        PolicyMap.Add(policyName, new DefaultRateLimiterPolicy(ConvertPartitioner<TPartitionKey>(partitioner), null));
+
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a new rate limiting policy with the given policyName.
+    /// </summary>
+    /// <param name="policyName">The name to be associated with the given TPolicy.</param>
+    public RateLimiterOptions AddPolicy<TPartitionKey, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPolicy>(string policyName) where TPolicy : IRateLimiterPolicy<TPartitionKey>
+    {
+        ArgumentNullException.ThrowIfNull(policyName);
+
+        if (PolicyMap.ContainsKey(policyName) || UnactivatedPolicyMap.ContainsKey(policyName))
+        {
+            throw new ArgumentException($"There already exists a policy with the name {policyName}");
+        }
+
+        var policyType = new PolicyTypeState(typeof(TPolicy));
```

Comment:
nit: move this into the lambda and mark the lambda as static
Unless that breaks the linker warning workaround

Reply:
I think it breaks it - @JamesNK?